### **Importando as bibliotecas**

Por enquanto, só será utilizado o pandas pra manipular os dados.

In [1]:
import pandas as pd

### **Transformando o arquivo .csv em um dataframe pandas**

In [2]:
tmdb_df = pd.read_csv('tmdb.csv')
tmdb_df.head()

id                                   title  \
0  1034541                             Terrifier 3   
1   912649                   Venom: The Last Dance   
2  1184918                          The Wild Robot   
3   933260                           The Substance   
4  1118031  Apocalypse Z: The Beginning of the End   

                                            overview release_date  runtime  \
0  Five years after surviving Art the Clown's Hal...   2024-10-09      125   
1  Eddie and Venom are on the run. Hunted by both...   2024-10-22      109   
2  After a shipwreck, an intelligent robot called...   2024-09-12      102   
3  A fading celebrity decides to use a black mark...   2024-09-07      141   
4  When a kind of rabies that transforms people i...   2024-10-04      119   

                               genres original_language     budget    revenue  \
0           Horror, Thriller, Mystery                en    2000000   78573405   
1  Science Fiction, Action, Adventure                en  120000000  394000000   
2  Animation, Science Fiction, Family                en   78000000  292583746   
3      Drama, Horror, Science Fiction                en   17500000   46818041   
4               Drama, Action, Horror                es          0          0   

   popularity  vote_average  vote_count  \
0    3304.003         6.900         886   
1    3930.216         6.385         658   
2    2214.674         8.500        2721   
3    1893.832         7.300        1809   
4    2835.786         6.800         389   

                                                cast  \
0  Lauren LaVera, Elliott Fullam, David Howard Th...   
1  Tom Hardy, Chiwetel Ejiofor, Juno Temple, Clar...   
2  Lupita Nyong'o, Pedro Pascal, Kit Connor, Bill...   
3  Demi Moore, Margaret Qualley, Dennis Quaid, Ed...   
4  Francisco Ortiz, José María Yázpik, Berta Vázq...   

                                                crew  \
0  Chris McGurk, Phil Falcone, G. Brandon Hill, B...   
1  David Michelinie, Todd McFarlane, Avi Arad, Ke...   
2  Dean DeBlois, Chris Sanders, Chris Sanders, Je...   
3  Matthieu de la Mortière, Anne Juin, Erik Baier...   
4  Adrián Guerra, Carles Torrens, Manel Loureiro,...   

                                production_companies  \
0  Cineverse, Bloody Disgusting, Dark Age Cinema,...   
1  Columbia Pictures, Pascal Pictures, Matt Tolma...   
2                               DreamWorks Animation   
3                    Working Title Films, Blacksmith   
4                                  Nostromo Pictures   

       production_countries                       poster_path  \
0  United States of America  /l1175hgL5DoXnqeZQCcU3eZIdhX.jpg   
1  United States of America  /aosm8NMQ3UyoBVpSxyimorCQykC.jpg   
2  United States of America  /wTnV3PCVW5O92JMrFvvrRcV39RU.jpg   
3    France, United Kingdom  /lqoMzCcZYEFK729d6qzt349fB4o.jpg   
4                     Spain  /wIGJnIFQlESkC2rLpfA8EDHqk4g.jpg   

                      backdrop_path  \
0  /18TSJF1WLA4CkymvVUcKDBwUJ9F.jpg   
1  /3V4kLQg0kSqPLctI5ziYWabAZYF.jpg   
2  /v9acaWVVFdZT5yAU7J2QjwfhXyD.jpg   
3  /7h6TqPB3ESmjuVbxCxAeB1c9OB1.jpg   
4  /2fxnTXr8NwyTFkunkimJkGkhqfy.jpg   

                                             tagline    status  
0          You better not shout, you better not cry.  Released  
1                           'Til death do they part.  Released  
2                         Discover your true nature.  Released  
3  If you follow the instructions, what could go ...  Released  
4                                                NaN  Released

In [3]:
tmdb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    14999 non-null  int64  
 1   title                 14999 non-null  object 
 2   overview              14923 non-null  object 
 3   release_date          14961 non-null  object 
 4   runtime               14999 non-null  int64  
 5   genres                14931 non-null  object 
 6   original_language     14999 non-null  object 
 7   budget                14999 non-null  int64  
 8   revenue               14999 non-null  int64  
 9   popularity            14999 non-null  float64
 10  vote_average          14999 non-null  float64
 11  vote_count            14999 non-null  int64  
 12  cast                  14912 non-null  object 
 13  crew                  14932 non-null  object 
 14  production_companies  14370 non-null  object 
 15  production_countrie

Aqui, é possível ver que há 14999 linhas, mas há algumas colunas com valores nulos.

### **Retirar os filmes que ainda não foram lançados**

Os filmes que não foram lançados, ou que estão como especulados, não servem para fazer o machine learning. Então, todas as linhas em que a coluna `status` esteja diferente de `Released` serão excluídas.

In [4]:
tmdb = tmdb_df[tmdb_df['status'] == 'Released']
tmdb.head()

id                                   title  \
0  1034541                             Terrifier 3   
1   912649                   Venom: The Last Dance   
2  1184918                          The Wild Robot   
3   933260                           The Substance   
4  1118031  Apocalypse Z: The Beginning of the End   

                                            overview release_date  runtime  \
0  Five years after surviving Art the Clown's Hal...   2024-10-09      125   
1  Eddie and Venom are on the run. Hunted by both...   2024-10-22      109   
2  After a shipwreck, an intelligent robot called...   2024-09-12      102   
3  A fading celebrity decides to use a black mark...   2024-09-07      141   
4  When a kind of rabies that transforms people i...   2024-10-04      119   

                               genres original_language     budget    revenue  \
0           Horror, Thriller, Mystery                en    2000000   78573405   
1  Science Fiction, Action, Adventure                en  120000000  394000000   
2  Animation, Science Fiction, Family                en   78000000  292583746   
3      Drama, Horror, Science Fiction                en   17500000   46818041   
4               Drama, Action, Horror                es          0          0   

   popularity  vote_average  vote_count  \
0    3304.003         6.900         886   
1    3930.216         6.385         658   
2    2214.674         8.500        2721   
3    1893.832         7.300        1809   
4    2835.786         6.800         389   

                                                cast  \
0  Lauren LaVera, Elliott Fullam, David Howard Th...   
1  Tom Hardy, Chiwetel Ejiofor, Juno Temple, Clar...   
2  Lupita Nyong'o, Pedro Pascal, Kit Connor, Bill...   
3  Demi Moore, Margaret Qualley, Dennis Quaid, Ed...   
4  Francisco Ortiz, José María Yázpik, Berta Vázq...   

                                                crew  \
0  Chris McGurk, Phil Falcone, G. Brandon Hill, B...   
1  David Michelinie, Todd McFarlane, Avi Arad, Ke...   
2  Dean DeBlois, Chris Sanders, Chris Sanders, Je...   
3  Matthieu de la Mortière, Anne Juin, Erik Baier...   
4  Adrián Guerra, Carles Torrens, Manel Loureiro,...   

                                production_companies  \
0  Cineverse, Bloody Disgusting, Dark Age Cinema,...   
1  Columbia Pictures, Pascal Pictures, Matt Tolma...   
2                               DreamWorks Animation   
3                    Working Title Films, Blacksmith   
4                                  Nostromo Pictures   

       production_countries                       poster_path  \
0  United States of America  /l1175hgL5DoXnqeZQCcU3eZIdhX.jpg   
1  United States of America  /aosm8NMQ3UyoBVpSxyimorCQykC.jpg   
2  United States of America  /wTnV3PCVW5O92JMrFvvrRcV39RU.jpg   
3    France, United Kingdom  /lqoMzCcZYEFK729d6qzt349fB4o.jpg   
4                     Spain  /wIGJnIFQlESkC2rLpfA8EDHqk4g.jpg   

                      backdrop_path  \
0  /18TSJF1WLA4CkymvVUcKDBwUJ9F.jpg   
1  /3V4kLQg0kSqPLctI5ziYWabAZYF.jpg   
2  /v9acaWVVFdZT5yAU7J2QjwfhXyD.jpg   
3  /7h6TqPB3ESmjuVbxCxAeB1c9OB1.jpg   
4  /2fxnTXr8NwyTFkunkimJkGkhqfy.jpg   

                                             tagline    status  
0          You better not shout, you better not cry.  Released  
1                           'Til death do they part.  Released  
2                         Discover your true nature.  Released  
3  If you follow the instructions, what could go ...  Released  
4                                                NaN  Released

In [5]:
tmdb.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14900 entries, 0 to 14998
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    14900 non-null  int64  
 1   title                 14900 non-null  object 
 2   overview              14831 non-null  object 
 3   release_date          14895 non-null  object 
 4   runtime               14900 non-null  int64  
 5   genres                14836 non-null  object 
 6   original_language     14900 non-null  object 
 7   budget                14900 non-null  int64  
 8   revenue               14900 non-null  int64  
 9   popularity            14900 non-null  float64
 10  vote_average          14900 non-null  float64
 11  vote_count            14900 non-null  int64  
 12  cast                  14819 non-null  object 
 13  crew                  14838 non-null  object 
 14  production_companies  14280 non-null  object 
 15  production_countries  14

### **Remover as colunas que não serão usadas**

Algumas colunas não serão o foco do projeto, ou serão muito trabalhosas para tratar.

In [6]:
tmdb = tmdb.drop(columns=['overview', 'poster_path', 'backdrop_path', 'cast', 'crew', 'tagline', 'status'])
tmdb.head()

id                                   title release_date  runtime  \
0  1034541                             Terrifier 3   2024-10-09      125   
1   912649                   Venom: The Last Dance   2024-10-22      109   
2  1184918                          The Wild Robot   2024-09-12      102   
3   933260                           The Substance   2024-09-07      141   
4  1118031  Apocalypse Z: The Beginning of the End   2024-10-04      119   

                               genres original_language     budget    revenue  \
0           Horror, Thriller, Mystery                en    2000000   78573405   
1  Science Fiction, Action, Adventure                en  120000000  394000000   
2  Animation, Science Fiction, Family                en   78000000  292583746   
3      Drama, Horror, Science Fiction                en   17500000   46818041   
4               Drama, Action, Horror                es          0          0   

   popularity  vote_average  vote_count  \
0    3304.003         6.900         886   
1    3930.216         6.385         658   
2    2214.674         8.500        2721   
3    1893.832         7.300        1809   
4    2835.786         6.800         389   

                                production_companies      production_countries  
0  Cineverse, Bloody Disgusting, Dark Age Cinema,...  United States of America  
1  Columbia Pictures, Pascal Pictures, Matt Tolma...  United States of America  
2                               DreamWorks Animation  United States of America  
3                    Working Title Films, Blacksmith    France, United Kingdom  
4                                  Nostromo Pictures                     Spain

In [7]:
tmdb.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14900 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    14900 non-null  int64  
 1   title                 14900 non-null  object 
 2   release_date          14895 non-null  object 
 3   runtime               14900 non-null  int64  
 4   genres                14836 non-null  object 
 5   original_language     14900 non-null  object 
 6   budget                14900 non-null  int64  
 7   revenue               14900 non-null  int64  
 8   popularity            14900 non-null  float64
 9   vote_average          14900 non-null  float64
 10  vote_count            14900 non-null  int64  
 11  production_companies  14280 non-null  object 
 12  production_countries  14646 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


### **Conferir quantos valores nulos existem em cada coluna**

In [8]:
print(tmdb.isna().sum())

id                        0
title                     0
release_date              5
runtime                   0
genres                   64
original_language         0
budget                    0
revenue                   0
popularity                0
vote_average              0
vote_count                0
production_companies    620
production_countries    254
dtype: int64


Todas as colunas que possuem dados nulos são importantes para o machine learning e os dados não podem ser inferidos através de outros, então precisam ser excluídos.

### **Remover todos os dados nulos**

In [9]:
tmdb_clean_df = tmdb.dropna()
tmdb_clean_df.tail()

id               title release_date  runtime  \
14994   13251          The Walker   2007-02-13      108   
14995   15265     The Red Balloon   1956-08-24       34   
14996   34027    The Great Gatsby   2000-03-29       90   
14997   52799            Undertow   2009-09-23      100   
14998  483104  A Christmas Prince   2017-11-17       92   

                               genres original_language    budget  revenue  \
14994                  Crime, Mystery                en  10000000   590416   
14995  Comedy, Drama, Family, Fantasy                fr         0    48980   
14996        Drama, Romance, TV Movie                en         0        0   
14997                  Drama, Romance                es         0        0   
14998               Romance, TV Movie                en         0        0   

       popularity  vote_average  vote_count  \
14994      18.371           5.2          95   
14995      14.020           7.7         393   
14996      12.620           5.5          20   
14997      10.521           7.4         118   
14998      15.525           6.6        1427   

                                    production_companies  \
14994  Kintop Pictures, Isle of Man Film, Paul Schrad...   
14995                                   Films Montsouris   
14996     Granada Entertainment, A&E Television Networks   
14997                             Dynamo, El Calvo Films   
14998              Motion Picture Corporation of America   

                                   production_countries  
14994          United Kingdom, United States of America  
14995                                            France  
14996  Canada, United Kingdom, United States of America  
14997                   Colombia, France, Germany, Peru  
14998                          United States of America

In [10]:
print(tmdb_clean_df.isna().sum())

id                      0
title                   0
release_date            0
runtime                 0
genres                  0
original_language       0
budget                  0
revenue                 0
popularity              0
vote_average            0
vote_count              0
production_companies    0
production_countries    0
dtype: int64


Agora, o dataframe não tem mais dados vazios.

### **Resetar o índice das linhas**

Como algumas linhas foram apagadas, o índice das linhas está com algumas lacunas que precisam ser ajeitadas.

In [11]:
tmdb_clean_df.reset_index(drop=True, inplace=True)
tmdb_clean_df.tail()

id               title release_date  runtime  \
14189   13251          The Walker   2007-02-13      108   
14190   15265     The Red Balloon   1956-08-24       34   
14191   34027    The Great Gatsby   2000-03-29       90   
14192   52799            Undertow   2009-09-23      100   
14193  483104  A Christmas Prince   2017-11-17       92   

                               genres original_language    budget  revenue  \
14189                  Crime, Mystery                en  10000000   590416   
14190  Comedy, Drama, Family, Fantasy                fr         0    48980   
14191        Drama, Romance, TV Movie                en         0        0   
14192                  Drama, Romance                es         0        0   
14193               Romance, TV Movie                en         0        0   

       popularity  vote_average  vote_count  \
14189      18.371           5.2          95   
14190      14.020           7.7         393   
14191      12.620           5.5          20   
14192      10.521           7.4         118   
14193      15.525           6.6        1427   

                                    production_companies  \
14189  Kintop Pictures, Isle of Man Film, Paul Schrad...   
14190                                   Films Montsouris   
14191     Granada Entertainment, A&E Television Networks   
14192                             Dynamo, El Calvo Films   
14193              Motion Picture Corporation of America   

                                   production_countries  
14189          United Kingdom, United States of America  
14190                                            France  
14191  Canada, United Kingdom, United States of America  
14192                   Colombia, France, Germany, Peru  
14193                          United States of America

In [12]:
tmdb_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14194 entries, 0 to 14193
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    14194 non-null  int64  
 1   title                 14194 non-null  object 
 2   release_date          14194 non-null  object 
 3   runtime               14194 non-null  int64  
 4   genres                14194 non-null  object 
 5   original_language     14194 non-null  object 
 6   budget                14194 non-null  int64  
 7   revenue               14194 non-null  int64  
 8   popularity            14194 non-null  float64
 9   vote_average          14194 non-null  float64
 10  vote_count            14194 non-null  int64  
 11  production_companies  14194 non-null  object 
 12  production_countries  14194 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 1.4+ MB


Após a limpeza, o dataframe que tinha 14999 linhas agora tem 14194, ou seja, foram excluídas 805 linhas.

### **Identificar os valores únicos das colunas de: gênero, língua original, empresa da produtora, país da produtora**

Esses dados podem ser transformados em números para trabalhar com os dados de uma forma mais objetiva.

In [13]:
tmdb_clean_df['genres'].drop_duplicates()

0                                Horror, Thriller, Mystery
1                       Science Fiction, Action, Adventure
2                       Animation, Science Fiction, Family
3                           Drama, Horror, Science Fiction
4                                    Drama, Action, Horror
                               ...                        
14141                                   Drama, Horror, War
14154    Animation, Comedy, Action, Family, Science Fic...
14169                  Comedy, Mystery, Thriller, TV Movie
14179            Animation, Comedy, Music, Family, Romance
14190                       Comedy, Drama, Family, Fantasy
Name: genres, Length: 2839, dtype: object

> Existem 2839 diferentes combinações de gêneros de filmes. Seria interessante pensar em algum jeito de catalogar cada gênero de forma única (ex.: horror = 0, thriller = 1, mystery = 2) e depois catalogar o gênero de cada filme de acordo com esses números (ex.: linha índice 0 -> 0, 1, 2).

In [14]:
tmdb_clean_df['original_language'].drop_duplicates()

0        en
4        es
11       fr
30       tl
31       ja
         ..
12277    mr
13031    yi
13174    mk
13686    kk
13811    si
Name: original_language, Length: 64, dtype: object

> Existem 64 tipos difentes de idiomas no dataframe. Essa coluna está mais fácil de traduzir as línguas para números.

In [15]:
tmdb_clean_df['production_companies'].drop_duplicates()

0        Cineverse, Bloody Disgusting, Dark Age Cinema,...
1        Columbia Pictures, Pascal Pictures, Matt Tolma...
2                                     DreamWorks Animation
3                          Working Title Films, Blacksmith
4                                        Nostromo Pictures
                               ...                        
14188             Sonar Entertainment, Lighthouse Pictures
14189    Kintop Pictures, Isle of Man Film, Paul Schrad...
14190                                     Films Montsouris
14191       Granada Entertainment, A&E Television Networks
14192                               Dynamo, El Calvo Films
Name: production_companies, Length: 11964, dtype: object

> Existem 11964 diferentes combinações de produtoras de filmes. É um número muito que talvez seja muito trabalhoso de tratar esses dados para eles ficarem mais  objetivos.

In [16]:
tmdb_clean_df['production_countries'].drop_duplicates()

0                                 United States of America
3                                   France, United Kingdom
4                                                    Spain
7                        Ireland, United States of America
11                                                  France
                               ...                        
14133             France, Georgia, Germany, United Kingdom
14141                        Jordan, Qatar, United Kingdom
14142    France, Monaco, Spain, Switzerland, United Sta...
14155         Argentina, Brazil, Chile, Netherlands, Spain
14192                      Colombia, France, Germany, Peru
Name: production_countries, Length: 1175, dtype: object

> Existem 1175 diferentes combinações de países de produtoras de filmes. É u número mais simples d se trabalhar do que o nome das empresas e seria interessante pensar em algum jeito de catalogar cada país de forma única (ex.: france = 1, united kingdom = 2) e depois catalogar o país de cada produtora de acordo com esses números (ex.: linha índice 3 -> 1, 2).

### **Atribuir um id a cada idioma**

In [17]:
idioma_original = tmdb_clean_df['original_language'].drop_duplicates().reset_index(drop=True)
language_id = {lang: idx for idx, lang in enumerate(idioma_original)}

tmdb_clean_df['original_language_id'] = tmdb_clean_df['original_language'].map(language_id)
tmdb_clean_df

<ipython-input-17-26a24b54ea7e>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmdb_clean_df['original_language_id'] = tmdb_clean_df['original_language'].map(language_id)


id                                   title release_date  runtime  \
0      1034541                             Terrifier 3   2024-10-09      125   
1       912649                   Venom: The Last Dance   2024-10-22      109   
2      1184918                          The Wild Robot   2024-09-12      102   
3       933260                           The Substance   2024-09-07      141   
4      1118031  Apocalypse Z: The Beginning of the End   2024-10-04      119   
...        ...                                     ...          ...      ...   
14189    13251                              The Walker   2007-02-13      108   
14190    15265                         The Red Balloon   1956-08-24       34   
14191    34027                        The Great Gatsby   2000-03-29       90   
14192    52799                                Undertow   2009-09-23      100   
14193   483104                      A Christmas Prince   2017-11-17       92   

                                   genres original_language     budget  \
0               Horror, Thriller, Mystery                en    2000000   
1      Science Fiction, Action, Adventure                en  120000000   
2      Animation, Science Fiction, Family                en   78000000   
3          Drama, Horror, Science Fiction                en   17500000   
4                   Drama, Action, Horror                es          0   
...                                   ...               ...        ...   
14189                      Crime, Mystery                en   10000000   
14190      Comedy, Drama, Family, Fantasy                fr          0   
14191            Drama, Romance, TV Movie                en          0   
14192                      Drama, Romance                es          0   
14193                   Romance, TV Movie                en          0   

         revenue  popularity  vote_average  vote_count  \
0       78573405    3304.003         6.900         886   
1      394000000    3930.216         6.385         658   
2      292583746    2214.674         8.500        2721   
3       46818041    1893.832         7.300        1809   
4              0    2835.786         6.800         389   
...          ...         ...           ...         ...   
14189     590416      18.371         5.200          95   
14190      48980      14.020         7.700         393   
14191          0      12.620         5.500          20   
14192          0      10.521         7.400         118   
14193          0      15.525         6.600        1427   

                                    production_companies  \
0      Cineverse, Bloody Disgusting, Dark Age Cinema,...   
1      Columbia Pictures, Pascal Pictures, Matt Tolma...   
2                                   DreamWorks Animation   
3                        Working Title Films, Blacksmith   
4                                      Nostromo Pictures   
...                                                  ...   
14189  Kintop Pictures, Isle of Man Film, Paul Schrad...   
14190                                   Films Montsouris   
14191     Granada Entertainment, A&E Television Networks   
14192                             Dynamo, El Calvo Films   
14193              Motion Picture Corporation of America   

                                   production_countries  original_language_id  
0                              United States of America                     0  
1                              United States of America                     0  
2                              United States of America                     0  
3                                France, United Kingdom                     0  
4                                                 Spain                     1  
...                                                 ...                   ...  
14189          United Kingdom, United States of America                     0  
14190                                            France                     2  
14191  Canada, United Ki

In [18]:
tmdb_clean_df['original_language_id'].drop_duplicates()

0         0
4         1
11        2
30        3
31        4
         ..
12277    59
13031    60
13174    61
13686    62
13811    63
Name: original_language_id, Length: 64, dtype: int64

### **Observações**

- Decidir se vai trabalhar, ou não, com as colunas `production_companies`, `cast`, `crew`.

- As colunas `budget` e `revenue` possuem dados não nulos, mas com o número 0, o que não faz muito sentido na situação, é preciso decidir qual o destino das linhas que estão assim.